core/src/test/java/org/apache/calcite/util/graph/DirectedGraphTest.java

```diff
@@ -172,6 +174,19 @@ private <V> List<List<V>> paths(DirectedGraph<V, DefaultEdge> g,
     assertThat(list, hasToString("[A, B, E, C, F, D]"));
   }
 
+  /**
+   * Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-7030">[CALCITE-7030]
+   * Enhance TopologicalOrderIterator to support BOTTOM_UP</a>. */
+  @Test void testTopologicalOrderWithBottomUpIterator() {
+    final DefaultDirectedGraph<String, DefaultEdge> graph = createDag();
+    final List<String> list = new ArrayList<>();
+    for (String s : TopologicalOrderIterator.of(graph, HepMatchOrder.BOTTOM_UP)) {
+      list.add(s);
+    }
+    assertThat(list, hasToString("[D, F, C, B, E, A]"));
```

Comment:
in general a graph has multiple topological orders, is the result here deterministic?

Reply:
Yes. It is deterministic. Because the data structures used during the traversal process are all in an ordered manner.